src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs

```diff
@@ -856,7 +869,13 @@ private void ThrowValidationException()
         static internal object UnsafeGetUninitializedObject(Type type)
         {
 #if !NET_NATIVE
-            return TryGetUninitializedObjectWithFormatterServices(type) ?? Activator.CreateInstance(type);
+            bool hasDefaultConstructor;
+            if (!_typeHasDefaultConstructorMap.TryGetValue(type, out hasDefaultConstructor))
+            {
+                hasDefaultConstructor = type.GetConstructor(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance, Array.Empty<Type>()) != null;
+                _typeHasDefaultConstructorMap[type] = hasDefaultConstructor;
+            }
+            return hasDefaultConstructor ? Activator.CreateInstance(type) : TryGetUninitializedObjectWithFormatterServices(type);
```

Comment:
The behavior of this method would be slightly different before/after the fix. Before the fix, if TryGetUninitializedObjectWithFormatterServices returned null and the type had no default ctor, Activator.CreateInstance would throw exception; after the fix, we would skip calling Activator.CreateInstance and UnsafeGetUninitializedObject would return null. Is this expected?